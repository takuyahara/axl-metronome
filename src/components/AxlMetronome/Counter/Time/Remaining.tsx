import React, { Component, createRef, RefObject } from 'react';
import _ from 'lodash';
import style from './Remaining.module.scss';
import Utils, { ICssSelector } from 'utils/Utils';

interface IProps {
  timeToCount: number;
  postProcess?: IPostProcess;
  inheritedSelector?: ICssSelector;
}
interface IState {
  remaining: number;
}
interface IPostProcess {
  onClick?: (e: React.MouseEvent) => void;
  onTouchEnd?: (e: React.TouchEvent) => void;
}
class Remaining extends Component<IProps, IState> {
  private postClick: (e: React.MouseEvent) => void;
  private postTouchEnd: (e: React.TouchEvent) => void;

  public constructor(props: IProps) {
    super(props);
    this.state = {
      remaining: props.timeToCount,
    };
    this.postClick = props.postProcess && props.postProcess.onClick ? props.postProcess.onClick : (e: React.MouseEvent) => {};
    this.postTouchEnd = props.postProcess && props.postProcess.onTouchEnd ? props.postProcess.onTouchEnd : (e: React.TouchEvent) => {};

    this.onClick = this.onClick.bind(this);
    this.onTouchEnd = this.onTouchEnd.bind(this);
  }
  private onClick(e: React.MouseEvent): void {
    this.postClick(e);
  }
  private onTouchEnd(e: React.TouchEvent): void {
    this.postTouchEnd(e);
  }
  private formatTime(timeInSecond: number): string
  {
    const minute = Math.floor(timeInSecond / 60);
    const second = Math.floor(timeInSecond % 60);
    const minuteStr = minute < 10 ? "0" + minute : "" + minute;
    const secondStr = second < 10 ? "0" + second : "" + second;
    return minuteStr + ":" + secondStr;
  }
  public setRemaining(remaining: number): void {
    this.setState({
      remaining,
    });
  }
  /* istanbul ignore next */
  public init(newProps: IProps): void {
    this.setState({
      remaining: newProps.timeToCount,
    });
  }
  /* istanbul ignore next */
  public componentWillMount(): void {
    if (this.props.inheritedSelector) {
      Utils.inheritSelector(style, this.props.inheritedSelector);
    }
  }
  /* istanbul ignore next */
  public componentWillReceiveProps(newProps: IProps): void {
    if (_.isEqual(this.props, newProps)) {
      return;
    }
    this.init(newProps);
  }
  public render(): React.ReactNode {
    const formattedTime = this.formatTime(Math.round(this.state.remaining));
    return (
      <div 
        className={style.indicator} 
        data-testid="indicator-remaining"
        onClick={this.onClick}
        onTouchEnd={this.onTouchEnd}
      >
        <div className={style.remaining}>
          {formattedTime}
        </div>
      </div>
    );
  }
}
export default Remaining;
