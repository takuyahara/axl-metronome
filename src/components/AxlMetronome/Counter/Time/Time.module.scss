@import "../../../mixins";
.time {
    width: 300px;
    height: 300px;
    @include centralize;
}
