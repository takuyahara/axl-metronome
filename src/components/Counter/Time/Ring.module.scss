@import "../../mixins";
$ringSize: 300px;
$ringStrokeWidth: 10px;
$ringCenter: $ringSize / 2;
$ringRadius: $ringSize / 2;
$maskStrokeWidth: 70px;

// Main
.ring {
    @include centralize;
    svg {
        width: $ringSize;
        height: $ringSize;
        user-select: none;
        -webkit-touch-callout: none;
        -webkit-user-select: none;
        -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
        .remaining {
            r: $ringRadius - $ringStrokeWidth / 2;
            cx: $ringCenter;
            cy: $ringCenter;
            fill: orange;
            stroke: orange;
            stroke-width: $ringStrokeWidth;
        }
        .mask {
            r: $ringRadius - $maskStrokeWidth / 2 - $ringStrokeWidth;
            cx: $ringCenter;
            cy: $ringCenter;
            fill: transparent;
            stroke-width: $maskStrokeWidth;
            transform: rotate(180deg);
            transform-origin: $ringCenter $ringCenter;
            opacity: 0.8;
        }
        .mask[stroke-dasharray^="0 "] {
            opacity: 0;
        }
        .tapArea {
            r: $ringRadius;
            cx: $ringCenter;
            cy: $ringCenter;
            fill: transparent;
            stroke: none;
        }
        .tapAreaMask {
            r: $ringRadius - $ringStrokeWidth - $maskStrokeWidth;
            cx: $ringCenter;
            cy: $ringCenter;
            fill: transparent;
            stroke: none;
        }
        #maskGradient stop[offset="30%"] {
            stop-color: rgb(60, 60, 60);
        }
        #maskGradient stop[offset="100%"] {
            stop-color: rgb(0, 0, 0);
        }
    }
    .globalTapArea[data-is-bubbled="false"] {
        display: none;
    }
    .globalTapArea[data-is-bubbled="true"] {
        display: block;
        position: absolute;
        top: calc(-50vh + #{$ringRadius});
        left: calc(-50vw + #{$ringRadius});
        width: 100vw;
        height: 100vh;
    }
}
