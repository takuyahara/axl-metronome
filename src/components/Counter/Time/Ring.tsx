import React, { Component, createRef, RefObject } from 'react';
import style from './Ring.module.scss';
import Utils, { ICssSelector } from 'utils/Utils';

export interface IProps {
  mouseEvent?: {
    onMouseMove: (e: React.MouseEvent | React.TouchEvent) => void;
    onMouseDown: (e: React.MouseEvent | React.TouchEvent) => void;
    onMouseUp: (e: React.MouseEvent | React.TouchEvent) => void;
  };
  inheritedSelector?: ICssSelector;
}
export interface IState {
  strokeDasharray: string;
  isTapped: boolean;
}
class Ring<P extends IProps = IProps, S extends IState = IState> extends Component<P, S> {
  // Initialize in init()
  protected progress: number = 0;
  protected postMouseMove!: (e: React.MouseEvent | React.TouchEvent) => void;
  protected postMouseDown!: (e: React.MouseEvent | React.TouchEvent) => void;
  protected postMouseUp!: (e: React.MouseEvent | React.TouchEvent) => void;

  // Ref
  protected refRing: RefObject<SVGSVGElement>;
  protected refMask: RefObject<SVGCircleElement>;

  public constructor(props: P) {
    super(props);
    this.state = {
      strokeDasharray: "",
      isTapped: false,
    } as S;

    // Ref
    this.refRing = createRef<SVGSVGElement>();
    this.refMask = createRef<SVGCircleElement>();

    // Post process
    this.postMouseMove = props.mouseEvent ? props.mouseEvent.onMouseMove : () => {};
    this.postMouseDown = props.mouseEvent ? props.mouseEvent.onMouseDown : () => {};
    this.postMouseUp = props.mouseEvent ? props.mouseEvent.onMouseUp : () => {};

    // Method binding
    this.mouseMove = this.mouseMove.bind(this);
    this.mouseUp = this.mouseUp.bind(this);
    this.mouseDown = this.mouseDown.bind(this);

    // Initialize
    this.progress = 0;
  }
  protected getBoundingClientRect(elem: Element): DOMRect {
    // Excluded to be replaced in test.
    return elem.getBoundingClientRect() as DOMRect;
  }
  protected getRatio(clientX: number, clientY: number): number {
    const refRing = this.refRing.current!;
    const ringGeometry = this.getBoundingClientRect(refRing);
    const centerX = ringGeometry.x + ringGeometry.width / 2;
    const centerY = ringGeometry.y + ringGeometry.height / 2;
    const dist = {
      x: clientX - centerX,
      y: clientY - centerY,
    };
    const radian = Math.atan2(dist.y, dist.x);
    const degree = radian * 180 / Math.PI;
    const ratio = (degree + 180) / 360;
    return ratio;
  }
  protected mouseMove(e: React.MouseEvent | React.TouchEvent): void {
    const isMouseEvent = e.type.substr(0, 5) === 'mouse';
    isMouseEvent && e.preventDefault();
    if (!this.state.isTapped) {
      return;
    }

    const clientX = isMouseEvent ? (e as React.MouseEvent).clientX : (e as React.TouchEvent).changedTouches[0].clientX;
    const clientY = isMouseEvent ? (e as React.MouseEvent).clientY : (e as React.TouchEvent).changedTouches[0].clientY;
    const progress = this.getRatio(clientX, clientY);
    this.setProgress(progress);
    this.postMouseMove(e);
  }
  protected mouseDown(e: React.MouseEvent | React.TouchEvent): void {
    const isMouseEvent = e.type.substr(0, 5) === 'mouse';
    isMouseEvent && e.preventDefault();
    
    const clientX = isMouseEvent ? (e as React.MouseEvent).clientX : (e as React.TouchEvent).changedTouches[0].clientX;
    const clientY = isMouseEvent ? (e as React.MouseEvent).clientY : (e as React.TouchEvent).changedTouches[0].clientY;
    const progress = this.getRatio(clientX, clientY);
    this.setProgress(progress);
    this.postMouseDown(e);
    this.setState({
      isTapped: true,
    });
  }
  protected mouseUp(e: React.MouseEvent | React.TouchEvent): void {
    const isMouseEvent = e.type.substr(0, 5) === 'mouse';
    isMouseEvent && e.preventDefault();

    this.postMouseUp(e);
    this.setState({
      isTapped: false,
    });
  }
  /* istanbul ignore next */
  protected getR(svgObject: SVGElement): number {
    return parseInt(getComputedStyle(svgObject).getPropertyValue('r'));
  }
  protected updateStrokeDasharray(): void {
    const refMask = this.refMask.current!;
    const r = this.getR(refMask);
    const pieLength = r * 2 * Math.PI;
    const strokeDasharray = this.progress * pieLength + " " + pieLength;
    this.setState({
      strokeDasharray,
    });
  }
  public reset(): void {
    this.setProgress(0.0);
  }
  public setProgress(progress: number): void {
    this.progress = progress;
    this.updateStrokeDasharray();
  }
  public getProgress(): number {
    return this.progress;
  }
  /* istanbul ignore next */
  // tslint:disable-next-line:no-any
  public init(...args: any): void {
    this.reset();
  }
  /* istanbul ignore next */
  public componentWillMount(): void {
    if (this.props.inheritedSelector) {
      Utils.inheritSelector(style, this.props.inheritedSelector!);
    }
  }
  /* istanbul ignore next */
  public componentDidMount(): void {
    this.updateStrokeDasharray();
  }
  public render(): React.ReactNode {
    return (
      <div 
        className={style.ring} 
      >
        <svg
          ref={this.refRing} 
        >
          <defs>
              <radialGradient gradientUnits="userSpaceOnUse" id="maskGradient" spreadMethod="reflect">
                <stop offset="30%" />
                <stop offset="100%" />
              </radialGradient>
          </defs>
          <circle 
            className={style.remaining} 
          />
          <circle ref={this.refMask} 
            className={style.mask} 
            stroke="url(#maskGradient)" 
            strokeDasharray={this.state.strokeDasharray} 
          />
          <circle 
            className={style.tapArea} 
            data-testid="ring-taparea" 
            onMouseDown={this.mouseDown} 
            onTouchStart={this.mouseDown} 
            onTouchMove={this.mouseMove} 
            onTouchEnd={this.mouseUp} 
          />
          <circle 
            className={style.tapAreaMask} 
          />
        </svg>
        <div 
          className={style.globalTapArea} 
          data-testid="ring-global-taparea" 
          data-is-bubbled={this.state.isTapped} 
          onMouseMove={this.mouseMove} 
          onMouseUp={this.mouseUp} 
        />
      </div>
    );
  }
}
export default Ring;