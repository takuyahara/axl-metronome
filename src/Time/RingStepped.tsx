import _ from 'lodash';
import Ring, { IProps as IPropsRing, IState as IStateRing } from './Ring';

interface IProps extends IPropsRing {
  step: number;
}
interface IState extends IStateRing {
  // nothing to extend
}
class RingStepped extends Ring<IProps, IState> {
  // Initialize in init()
  private step!: number;

  public constructor(props: IProps) {
    super(props);
    this.init(props);
  }
  protected init(props: IProps): void {
    this.step = props.step;
    super.init(props);
  }
  public setProgress(progress: number): void {
    super.setProgress(Math.floor((progress * this.step)) / this.step); // FIXME: Rounding Error
  }
  /* istanbul ignore next */
  public componentWillReceiveProps(newProps: IProps): void {
    if (_.isEqual(this.props, newProps)) {
      return;
    }

    this.init(newProps);
    this.reset();
  }
}
export default RingStepped;