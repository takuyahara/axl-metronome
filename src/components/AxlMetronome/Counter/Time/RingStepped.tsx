import _ from 'lodash';
import Ring, { IProps as IPropsRing, IState as IStateRing } from './Ring';

interface IProps extends IPropsRing {
  step: number;
}
interface IState extends IStateRing {
  // nothing to extend
}
class RingStepped extends Ring<IProps, IState> {
  private step: number;

  public constructor(props: IProps) {
    super(props);
    this.step = props.step;
  }
  protected mouseMove(e: React.MouseEvent | React.TouchEvent): void {
    const isMouseEvent = e.type.substr(0, 5) === 'mouse';
    isMouseEvent && e.preventDefault();
    if (!this.isTapped) {
      return;
    }

    const clientX = isMouseEvent ? (e as React.MouseEvent).clientX : (e as React.TouchEvent).changedTouches[0].clientX;
    const clientY = isMouseEvent ? (e as React.MouseEvent).clientY : (e as React.TouchEvent).changedTouches[0].clientY;
    const progress = this.getRatio(clientX, clientY);
    const progressStepped = Math.floor((progress * this.step)) / this.step;
    this.setProgress(progressStepped);
    this.postMouseMove(e);
  }
  protected mouseDown(e: React.MouseEvent | React.TouchEvent): void {
    const isMouseEvent = e.type.substr(0, 5) === 'mouse';
    isMouseEvent && e.preventDefault();
    
    const clientX = isMouseEvent ? (e as React.MouseEvent).clientX : (e as React.TouchEvent).changedTouches[0].clientX;
    const clientY = isMouseEvent ? (e as React.MouseEvent).clientY : (e as React.TouchEvent).changedTouches[0].clientY;
    const progress = this.getRatio(clientX, clientY);
    const progressStepped = Math.floor((progress * this.step)) / this.step;
    this.setProgress(progressStepped);
    this.postMouseDown(e);
    this.isTapped = true;
    if (isMouseEvent) {
      this.setState({
        isBubbled: true,
      });
    }
  }
  /* istanbul ignore next */
  public init(props: IProps): void {
    super.init();
    this.step = props.step;
  }
  /* istanbul ignore next */
  public componentWillReceiveProps(newProps: IProps): void {
    if (_.isEqual(this.props, newProps)) {
      return;
    }

    this.init(newProps);
    this.reset();
  }
}
export default RingStepped;