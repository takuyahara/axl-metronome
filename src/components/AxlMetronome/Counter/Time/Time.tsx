import React, { Component, createRef, RefObject } from 'react';
import _ from 'lodash';
import Indicator from './Indicator';
import { default as Ring } from './RingStepped';
import style from './Time.module.scss';
import Status from './Status';
import Utils, { ICssSelector } from 'utils/Utils';

interface IProps {
  remaining: number;
  postProcess?: {
    Ring?: {
      onMouseDown?: (e: React.MouseEvent | React.TouchEvent) => void;
      onMouseMove?: (e: React.MouseEvent | React.TouchEvent) => void;
      onMouseUp?: (e: React.MouseEvent | React.TouchEvent) => void;
    };
    Indicator?: {
      onCount?: () => void;
      onCountStart?: () => void;
      onCountStop?: () => void;
      onCountPause?: () => void;
      onCountResume?: () => void;
      onCountComplete?: () => void;
    };
  };
  inheritedSelector?: ICssSelector;
}
interface IState {
  // empty
}
class Time extends Component<IProps, IState> {
  private postProcess: {
    Ring: {
      onMouseDown: (e: React.MouseEvent | React.TouchEvent) => void;
      onMouseMove: (e: React.MouseEvent | React.TouchEvent) => void;
      onMouseUp: (e: React.MouseEvent | React.TouchEvent) => void;
    };
    Indicator: {
      onCount: () => void;
      onCountStart: () => void;
      onCountStop: () => void;
      onCountPause: () => void;
      onCountResume: () => void;
      onCountComplete: () => void;
    };
  };
  private isDragging: boolean = false;
  private refIndicator: RefObject<Indicator>;
  private refRing: RefObject<Ring>;

  public constructor(props: IProps) {
    super(props);
    // this.state = {
    //
    // };

    // Ref
    this.refIndicator = createRef<Indicator>();
    this.refRing = createRef<Ring>();

    // Post process
    this.postProcess = {
      Ring: {
        onMouseDown: props.postProcess && props.postProcess.Ring && props.postProcess.Ring.onMouseDown ? 
          props.postProcess.Ring.onMouseDown : (e: React.MouseEvent | React.TouchEvent) => {},
        onMouseMove: props.postProcess && props.postProcess.Ring && props.postProcess.Ring.onMouseMove ? 
          props.postProcess.Ring.onMouseMove : (e: React.MouseEvent | React.TouchEvent) => {},
        onMouseUp: props.postProcess && props.postProcess.Ring && props.postProcess.Ring.onMouseUp ? 
          props.postProcess.Ring.onMouseUp : (e: React.MouseEvent | React.TouchEvent) => {},
      },
      Indicator: {
        onCount: props.postProcess && props.postProcess.Indicator && props.postProcess.Indicator.onCount ? 
          props.postProcess.Indicator.onCount : () => {},
        onCountStart: props.postProcess && props.postProcess.Indicator && props.postProcess.Indicator.onCountStart ? 
          props.postProcess.Indicator.onCountStart : () => {},
        onCountStop: props.postProcess && props.postProcess.Indicator && props.postProcess.Indicator.onCountStop ? 
          props.postProcess.Indicator.onCountStop : () => {},
        onCountPause: props.postProcess && props.postProcess.Indicator && props.postProcess.Indicator.onCountPause ? 
          props.postProcess.Indicator.onCountPause : () => {},
        onCountResume: props.postProcess && props.postProcess.Indicator && props.postProcess.Indicator.onCountResume ? 
          props.postProcess.Indicator.onCountResume : () => {},
        onCountComplete: props.postProcess && props.postProcess.Indicator && props.postProcess.Indicator.onCountComplete ? 
          props.postProcess.Indicator.onCountComplete : () => {},
      },
    };

    // Method binding
    this.onRingMouseDown = this.onRingMouseDown.bind(this);
    this.onRingMouseMove = this.onRingMouseMove.bind(this);
    this.onRingMouseUp = this.onRingMouseUp.bind(this);
    this.onIndicatorCount = this.onIndicatorCount.bind(this);
    this.onIndicatorCountStart = this.onIndicatorCountStart.bind(this);
    this.onIndicatorCountStop = this.onIndicatorCountStop.bind(this);
    this.onIndicatorCountPause = this.onIndicatorCountPause.bind(this);
    this.onIndicatorCountResume = this.onIndicatorCountResume.bind(this);
    this.onIndicatorCountComplete = this.onIndicatorCountComplete.bind(this);
  }
  private onRingMouseDown(e: React.MouseEvent | React.TouchEvent): void {
    const refRing = this.refRing.current!;
    const refIndicator = this.refIndicator.current!;
    const progress = refRing.getProgress();
    this.setProgress(progress);
    if (refIndicator.getState() === Status.Running) {
      refIndicator.pause();
      this.isDragging = true;
    }
    this.postProcess.Ring.onMouseDown(e);
  }
  private onRingMouseMove(e: React.MouseEvent | React.TouchEvent): void {
    const refRing = this.refRing.current!;
    const progress = refRing.getProgress();
    this.setProgress(progress);
    this.postProcess.Ring.onMouseMove(e);
  }
  private onRingMouseUp(e: React.MouseEvent | React.TouchEvent): void {
    const refIndicator = this.refIndicator.current!;
    if (this.isDragging) {
      refIndicator.resume();
      this.isDragging = false;
    }
    this.postProcess.Ring.onMouseUp(e);
  }
  private onIndicatorCount(): void {
    const refRing = this.refRing.current!;
    const refIndicator = this.refIndicator.current!;
    refRing.setProgress(refIndicator.getProgress());
    this.postProcess.Indicator.onCount();
  }
  private onIndicatorCountStart(): void {
    this.postProcess.Indicator.onCountStart();
  }
  private onIndicatorCountStop(): void {
    this.postProcess.Indicator.onCountStop();
  }
  private onIndicatorCountPause(): void {
    this.postProcess.Indicator.onCountPause();
  }
  private onIndicatorCountResume(): void {
    this.postProcess.Indicator.onCountResume();
  }
  private onIndicatorCountComplete(): void {
    const refRing = this.refRing.current!;
    refRing.reset();
    this.postProcess.Indicator.onCountComplete();
  }
  public start(): void {
    const refIndicator = this.refIndicator.current!;
    refIndicator.start();
  }
  public stop(): void {
    const refIndicator = this.refIndicator.current!;
    const refRing = this.refRing.current!;
    refIndicator.stop();
    refRing.reset();
  }
  public pause(): void {
    const refIndicator = this.refIndicator.current!;
    refIndicator.pause();
  }
  public resume(): void {
    const refIndicator = this.refIndicator.current!;
    refIndicator.resume();
  }
  public setProgress(progress: number): void {
    const refRing = this.refRing.current!;
    const refIndicator = this.refIndicator.current!;
    refRing.setProgress(progress);
    refIndicator.setProgress(progress);
  }
  public getProgress(): number {
    const refRing = this.refRing.current!;
    return refRing.getProgress();
  }
  public getState(): Status {
    const refIndicator = this.refIndicator.current!;
    return refIndicator.getState();
  }
  /* istanbul ignore next */
  public init(newProps: IProps): void {
    const refRing = this.refRing.current!;
    const refIndicator = this.refIndicator.current!;
    refRing.init({
      step: newProps.remaining,
    });
    refIndicator.init({
      timeToCount: newProps.remaining,
    });
  }
  /* istanbul ignore next */
  public componentWillMount(): void {
    if (this.props.inheritedSelector) {
      Utils.inheritSelector(style, this.props.inheritedSelector!);
    }
  }
  /* istanbul ignore next */
  public componentWillReceiveProps(newProps: IProps): void {
    if (_.isEqual(this.props, newProps)) {
      return;
    }
    this.init(newProps);
  }
  public render(): React.ReactNode {
    return (
      <div 
        className={style.time} 
      >
        <Ring 
          ref={this.refRing} 
          step={this.props.remaining}
          mouseEvent={{
            onMouseMove: this.onRingMouseMove,
            onMouseDown: this.onRingMouseDown,
            onMouseUp: this.onRingMouseUp,
          }} 
          inheritedSelector={style} 
        />
        <Indicator 
          ref={this.refIndicator} 
          timeToCount={this.props.remaining} 
          postProcess={{
            onCount: this.onIndicatorCount,
            onCountStart: this.onIndicatorCountStart,
            onCountStop: this.onIndicatorCountStop,
            onCountPause: this.onIndicatorCountPause,
            onCountResume: this.onIndicatorCountResume,
            onCountComplete: this.onIndicatorCountComplete,
          }}
          inheritedSelector={style} 
        />
      </div>
    );
  }
}
export default Time;