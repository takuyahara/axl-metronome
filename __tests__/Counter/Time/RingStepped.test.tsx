import React from 'react';
import { configure, mount, ReactWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import sinon from 'sinon';
import { default as Ring } from 'AxlMetronome/Counter/Time/RingStepped';

configure({ adapter: new Adapter() });
const sel = (id: string) => `[data-testid="${id}"]`;

// tslint:disable:no-any
describe('RingStepped', () => {
  let ring: ReactWrapper;
  let ringInstance: Ring;
  let spy: { [s: string]: sinon.SinonSpy };

  beforeEach(() => {
    // Sets handlers before mounting.
    spy = {
      setProgress: sinon.spy(Ring.prototype, 'setProgress'),
      updateStrokeDasharray: sinon.spy(Ring.prototype as any, 'updateStrokeDasharray'),
    };
    sinon.replace(Ring.prototype as any, 'getR', () => 145);
    sinon.replace(Element.prototype as any, 'getBoundingClientRect', () => ({
      left: 0, 
      top: 0, 
      width: 300, 
      height: 300, 
    }));

    // Mounts VDOM and makes instance.
    ring = mount(<Ring
      step={5}
    />);
    ringInstance = ring.instance() as Ring;

    sinon.resetHistory();
  });
  afterEach(() => {
    ring.unmount();
    sinon.restore();
  });

  describe('inherited functons', () => {
    it('returns progress', () => {
      expect(ringInstance.getProgress()).toBe(0.0);
    });
  
    it('changes progress', () => {
      ringInstance.setProgress(0.5);
      expect(spy.updateStrokeDasharray.calledOnce).toBe(true);
      expect(ringInstance.getProgress()).toBe(0.5);
    });
  
    it('resets progress', () => {
      ringInstance.setProgress(0.875);
      expect(spy.updateStrokeDasharray.calledOnce).toBe(true);
  
      ringInstance.reset();
      expect(spy.updateStrokeDasharray.calledTwice).toBe(true);
      expect(ringInstance.getProgress()).toBe(0.0);
  
    });
  
    it('gets ratio', () => {
      expect((ringInstance as any).getRatio(150, 0)).toBe(0.25);
      expect((ringInstance as any).getRatio(300, 150)).toBe(0.5);
      expect((ringInstance as any).getRatio(150, 300)).toBe(0.75);
    });
  });

  describe('change value by MouseEvent', () => {
    afterEach(() => {
      ring.update();
      expect(ring).toMatchSnapshot();
    });

    it('mouseMove', () => {
      ring.find(sel('ring-global-taparea')).simulate('mousemove', {
        clientX: 300,
        clientY: 150,
      });

      expect(spy.setProgress.notCalled).toBe(true);
      expect(ringInstance.getProgress()).toBe(0.0);
    });

    it('mouseDown', () => {
      ring.find(sel('ring-taparea')).simulate('mousedown', {
        clientX: 300,
        clientY: 150,
      });

      expect(spy.setProgress.calledOnce).toBe(true);
      expect(ringInstance.getProgress()).toBe(0.4);
    });

    it('mouseDown => mouseMove', () => {
      ring.find(sel('ring-taparea')).simulate('mousedown', {
        clientX: 300,
        clientY: 150,
      });

      ring.find(sel('ring-global-taparea')).simulate('mousemove', {
        clientX: 150,
        clientY: 0,
      });

      expect(spy.setProgress.calledTwice).toBe(true);
      expect(ringInstance.getProgress()).toBe(0.2);
    });

    it('mouseDown => mouseMove => mouseUp', () => {
      ring.find(sel('ring-taparea')).simulate('mousedown', {
        clientX: 300,
        clientY: 150,
      });

      ring.find(sel('ring-global-taparea')).simulate('mousemove', {
        clientX: 150,
        clientY: 0,
      });

      ring.find(sel('ring-global-taparea')).simulate('mouseup');

      expect(spy.setProgress.calledTwice).toBe(true);
      expect(ringInstance.getProgress()).toBe(0.2);
    });
  });

  describe('change value by TouchEvent', () => {
    it('touchmove', () => {
      ring.find(sel('ring-global-taparea')).simulate('touchmove', {
        changedTouches: [{
          clientX: 300,
          clientY: 150,
        }],
      });

      expect(spy.setProgress.notCalled).toBe(true);
      expect(ringInstance.getProgress()).toBe(0.0);
    });

    it('touchstart', () => {
      ring.find(sel('ring-taparea')).simulate('touchstart', {
        changedTouches: [{
          clientX: 300,
          clientY: 150,
        }],
      });

      expect(spy.setProgress.calledOnce).toBe(true);
      expect(ringInstance.getProgress()).toBe(0.4);
    });

    it('touchstart => touchmove', () => {
      ring.find(sel('ring-taparea')).simulate('touchstart', {
        changedTouches: [{
          clientX: 300,
          clientY: 150,
        }],
      });

      ring.find(sel('ring-taparea')).simulate('touchmove', {
        changedTouches: [{
          clientX: 150,
          clientY: 0,
        }],
      });

      expect(spy.setProgress.calledTwice).toBe(true);
      expect(ringInstance.getProgress()).toBe(0.2);
    });

    it('touchstart => touchmove => touchend', () => {
      ring.find(sel('ring-taparea')).simulate('touchstart', {
        changedTouches: [{
          clientX: 300,
          clientY: 150,
        }],
      });

      ring.find(sel('ring-taparea')).simulate('touchmove', {
        changedTouches: [{
          clientX: 150,
          clientY: 0,
        }],
      });

      ring.find(sel('ring-taparea')).simulate('touchend');

      expect(spy.setProgress.calledTwice).toBe(true);
      expect(ringInstance.getProgress()).toBe(0.2);
    });
  });
});
