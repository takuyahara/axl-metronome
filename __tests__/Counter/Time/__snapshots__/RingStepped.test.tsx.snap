// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RingStepped change value by MouseEvent mouseDown => mouseMove => mouseUp 1`] = `
<RingStepped
  step={5}
>
  <div
    className="ring"
  >
    <svg>
      <defs>
        <radialGradient
          gradientUnits="userSpaceOnUse"
          id="maskGradient"
          spreadMethod="reflect"
        >
          <stop
            offset="30%"
          />
          <stop
            offset="100%"
          />
        </radialGradient>
      </defs>
      <circle
        className="remaining"
      />
      <circle
        className="mask"
        stroke="url(#maskGradient)"
        strokeDasharray="182.212373908208 911.06186954104"
      />
      <circle
        className="tapArea"
        data-testid="ring-taparea"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      />
      <circle
        className="tapAreaMask"
      />
    </svg>
    <div
      className="globalTapArea"
      data-is-bubbled={false}
      data-testid="ring-global-taparea"
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
    />
  </div>
</RingStepped>
`;

exports[`RingStepped change value by MouseEvent mouseDown => mouseMove 1`] = `
<RingStepped
  step={5}
>
  <div
    className="ring"
  >
    <svg>
      <defs>
        <radialGradient
          gradientUnits="userSpaceOnUse"
          id="maskGradient"
          spreadMethod="reflect"
        >
          <stop
            offset="30%"
          />
          <stop
            offset="100%"
          />
        </radialGradient>
      </defs>
      <circle
        className="remaining"
      />
      <circle
        className="mask"
        stroke="url(#maskGradient)"
        strokeDasharray="182.212373908208 911.06186954104"
      />
      <circle
        className="tapArea"
        data-testid="ring-taparea"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      />
      <circle
        className="tapAreaMask"
      />
    </svg>
    <div
      className="globalTapArea"
      data-is-bubbled={true}
      data-testid="ring-global-taparea"
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
    />
  </div>
</RingStepped>
`;

exports[`RingStepped change value by MouseEvent mouseDown 1`] = `
<RingStepped
  step={5}
>
  <div
    className="ring"
  >
    <svg>
      <defs>
        <radialGradient
          gradientUnits="userSpaceOnUse"
          id="maskGradient"
          spreadMethod="reflect"
        >
          <stop
            offset="30%"
          />
          <stop
            offset="100%"
          />
        </radialGradient>
      </defs>
      <circle
        className="remaining"
      />
      <circle
        className="mask"
        stroke="url(#maskGradient)"
        strokeDasharray="364.424747816416 911.06186954104"
      />
      <circle
        className="tapArea"
        data-testid="ring-taparea"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      />
      <circle
        className="tapAreaMask"
      />
    </svg>
    <div
      className="globalTapArea"
      data-is-bubbled={true}
      data-testid="ring-global-taparea"
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
    />
  </div>
</RingStepped>
`;

exports[`RingStepped change value by MouseEvent mouseMove 1`] = `
<RingStepped
  step={5}
>
  <div
    className="ring"
  >
    <svg>
      <defs>
        <radialGradient
          gradientUnits="userSpaceOnUse"
          id="maskGradient"
          spreadMethod="reflect"
        >
          <stop
            offset="30%"
          />
          <stop
            offset="100%"
          />
        </radialGradient>
      </defs>
      <circle
        className="remaining"
      />
      <circle
        className="mask"
        stroke="url(#maskGradient)"
        strokeDasharray="0 911.06186954104"
      />
      <circle
        className="tapArea"
        data-testid="ring-taparea"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      />
      <circle
        className="tapAreaMask"
      />
    </svg>
    <div
      className="globalTapArea"
      data-is-bubbled={false}
      data-testid="ring-global-taparea"
      onMouseMove={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
    />
  </div>
</RingStepped>
`;
